presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -134,6 +157,12 @@ private boolean exploreNode(int group, Context context)
         return progress;
     }
 
+    // It's not a hard limit due to inaccuracy of time registered in ruleStats
+    private boolean isTimeLimitExhausted()
+    {
+        return stats.getApproxTotalTimeInMillis() >= timeLimitInMillis;
```

Comment:
Instead of measuring time by relying on what's tracked in `stats`, just record the start time when `optimize` is called (store it in the Context object) and check against that.